baseplate/lib/circuit_breaker/factory.py

```diff
@@ -0,0 +1,164 @@
+import logging
+
+from contextlib import contextmanager
+from datetime import timedelta
+from typing import Any
+from typing import Dict
+from typing import Iterator
+from typing import Tuple
+from typing import Type
+
+from baseplate import Span
+from baseplate.clients import ContextFactory
+from baseplate.lib import config
+from baseplate.lib.circuit_breaker.breaker import Breaker
+from baseplate.lib.circuit_breaker.breaker import BreakerState
+from baseplate.lib.circuit_breaker.errors import BreakerTrippedError
+
+
+logger = logging.getLogger(__name__)
+
+
+class CircuitBreakerClientWrapperFactory(ContextFactory):
+    """Provide an object combining a client and circuit breaker for use with the client.
+
+    When attached to the baseplate `RequestContext` can be used like:
+
+    ```
+    breakable_exceptions = (...)  # exceptions indicating the service is unhealthy
+    with context.breaker_wrapped_client.breaker_context("identifier", breakable_exceptions) as svc:
+        svc.get_something()
+    ```
+    """
+
+    def __init__(self, client_factory: ContextFactory, breaker_box: "CircuitBreakerBox"):
+        self.client_factory = client_factory
+        self.breaker_box = breaker_box
+
+    def make_object_for_context(self, name: str, span: Span) -> Any:
+        client = self.client_factory.make_object_for_context(name, span)
+
+        return CircuitBreakerWrappedClient(span, self.breaker_box, client)
+
+
+class CircuitBreakerWrappedClient:
```

Comment:
Either way may be ideally out of scope for this PR but it could be nice to have a `ThriftCircuitBreakerWrappedClient` that acts like a proxy class and just decorates the thrift interface functions with the `breaker_context` logic.

Reply:
I wrote a little implementation in our codebase for this, works fine. I wonder if a patch to integrate a circuit breaker with the thrift client would be accepted after this merges.  Since retry logic can cause N requests to be sent before it lets the last exception bubble up to the breaker doesn't really have an accurate picture.